osu.Game.Rulesets.Osu/Edit/Blueprints/Sliders/Components/SliderTailPiece.cs

```diff
@@ -0,0 +1,210 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Caching;
+using osu.Framework.Extensions.Color4Extensions;
+using osu.Framework.Input;
+using osu.Framework.Input.Events;
+using osu.Framework.Utils;
+using osu.Game.Graphics;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Osu.Objects;
+using osu.Game.Screens.Edit;
+using osuTK;
+using osuTK.Graphics;
+using osuTK.Input;
+
+namespace osu.Game.Rulesets.Osu.Edit.Blueprints.Sliders.Components
+{
+    public partial class SliderTailPiece : SliderCircleOverlay
+    {
+        /// <summary>
+        /// Whether this is currently being dragged.
+        /// </summary>
+        private bool isDragging;
+
+        private InputManager inputManager = null!;
+
+        private readonly Cached<SliderPath> fullPathCache = new Cached<SliderPath>();
+
+        [Resolved(CanBeNull = true)]
+        private EditorBeatmap? editorBeatmap { get; set; }
+
+        [Resolved]
+        private OsuColour colours { get; set; } = null!;
+
+        public SliderTailPiece(Slider slider, SliderPosition position)
+            : base(slider, position)
+        {
+            Slider.Path.ControlPoints.CollectionChanged += (_, _) => fullPathCache.Invalidate();
+        }
+
+        protected override void LoadComplete()
+        {
+            base.LoadComplete();
+
+            inputManager = GetContainingInputManager();
+        }
+
+        public override bool ReceivePositionalInputAt(Vector2 screenSpacePos) => CirclePiece.ReceivePositionalInputAt(screenSpacePos);
+
+        protected override bool OnHover(HoverEvent e)
+        {
+            updateCirclePieceColour();
+            return false;
+        }
+
+        protected override void OnHoverLost(HoverLostEvent e)
+        {
+            updateCirclePieceColour();
+        }
+
+        protected override bool OnKeyDown(KeyDownEvent e)
+        {
+            if (e.Repeat)
+                return false;
+
+            handleDragToggle(e);
+            return base.OnKeyDown(e);
+        }
+
+        protected override void OnKeyUp(KeyUpEvent e)
+        {
+            handleDragToggle(e);
+            base.OnKeyUp(e);
+        }
+
+        private bool lastShiftPressed;
+
+        private void handleDragToggle(KeyboardEvent key)
+        {
+            bool shiftPressed = key.ShiftPressed;
+
+            if (shiftPressed == lastShiftPressed) return;
+
+            lastShiftPressed = shiftPressed;
```

Comment:
What is all of this `shiftPressed` / `lastShiftPressed` tracking for? It doesn't really look like it's doing much useful to me.

Reply:
It is to make sure `updateCirclePieceColour` only triggers on changes in the shift key state. I'll simplify the logic.